### ISSUE TITLE
Guidance Needed on Implementing Stacked Observations for Non-Image Data in Custom Environment

### ISSUE THREAD

**Issue by beuricav** [DLR-RM/stable-baselines3#1800]
### 📚 Documentation

I am currently working on integrating the built-in stacked observations feature of Stable Baselines 3 into my custom environment, but I'm facing challenges in applying it correctly.

**Context:**

- Objective: I aim to provide "stacked observations" as a 2D matrix to an RL agent.

- Reason: My environment already has around 340 input variables. Flattening a stack of 2 would excessively increase the input size, impacting the neural network size and complexity.

- Current Implementation: I've manually created stacked observations within my environment, resulting in 2D observations being returned. While this approach works, and the code runs with SB3's PPO and recurrent PPO (SB3-Contrib), I receive a check_env warning about unconventional observation shapes.

- Concern: I'm not confident that the SB3 policies (e.g., MlpPolicy) are processing the 2D stacked observations effectively. It seems they might only be considering the first observation in the stack.

- Attempt with VecFrameStack: When trying to use VecFrameStack, it recognizes that I'm not dealing with image data and attempts to use the StackedObservations wrapper. However, this leads to issues, particularly when integrating with the EvalCallback in conjunction with a stacked eval_env, resulting in errors I've been unable to resolve.

**Assumption:**

- My assumption is that my custom environment, which is compatible with DummyVecEnv, might not be fully compatible with VecFrameStack.

**Request:**

- I am seeking detailed documentation or guidelines on correctly applying stacked observations for non-image data.
- Are there any examples or best practices for implementing this feature in environments with a large number of input features?
- How can I ensure that the policies in SB3 are effectively utilizing the full stack of observations?

**Additional Notes:**

I am aware of the Recurrent PPO from SB3-Contrib. However, due to its slow performance, I am currently exploring stacking as an alternative for faster experimentation.

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)

**Comment by beuricav** [comment]
To Add to my Request. Here a bief information about my RL Problem:

- I like to provide "stacked observations" as a 2D matrix to the agent. Reason being I already have 340 inputs variables flatten out the a stack = 2  would increases the NN even more
- Current I could manually create the stackObservations in my environment it self, so it return 2D Observations
- When doing it directly using my own code, I of cause get the check_env warning that I have an unusal Observation but the code "runs", but I don't believe the Policy is using truly using the stacked data. 
- I was not able to find a hing in the source code if the PPO, MLPpPolicy is truly processing the 2D observations, or if it simply only checks the first obsevation and not the stacked 2nd. 
- I tried to use VecFrameStack and the wrapper realises I don't have image data and tries to utilize the "stackObsevations" wrapper, because it asks for the inputs varaibles of it in my environment. 
- Using then the VecFrameStack I get my model to train, but when I then use the EvalCallback with my as well stacked eval_env I get a error I cannot resolve. 
- My current assumption is either my custom environment, which was compatible with DummyVecenv, is not compatible with VecFramestack 


Therefore I try to find a more detailed documentation on how to properly apply stacked observations for non image data. 

PS: of cause I'm aware of RecurrentPPO of the SB3-Contrib, but the performance is very slow and currently I rather prefer stacking which allows much faster experimentation.

**Comment by beuricav** [comment]
Ok I digged deeper in the source code and documentation and found finally the hint, I can print my model policy. Doing this I see, that in the preprocessing the inputs are flatten. So even you get a warning the stacked inputs get flatten.  So stacking is still the way to go, but I need to define a custom policy network to cope with the increased input nodes. 

/opt/homebrew/Caskroom/miniconda/base/envs/py311/lib/python3.11/site-packages/stable_baselines3/common/env_checker.py:251: UserWarning: Your observation  has an unconventional shape (neither an image, nor a 1D vector). We recommend you to flatten the observation to have only a 1D vector or use a custom policy to properly process the data.
  warnings.warn(
ActorCriticPolicy(
  (features_extractor): FlattenExtractor(
    (flatten): Flatten(start_dim=1, end_dim=-1)
  )
  (pi_features_extractor): FlattenExtractor(
    (flatten): Flatten(start_dim=1, end_dim=-1)
  )
  (vf_features_extractor): FlattenExtractor(
    (flatten): Flatten(start_dim=1, end_dim=-1)
  )
  (mlp_extractor): MlpExtractor(
    (policy_net): Sequential(
      (0): Linear(in_features=2901, out_features=64, bias=True)
      (1): Tanh()
      (2): Linear(in_features=64, out_features=64, bias=True)
      (3): Tanh()
    )
    (value_net): Sequential(
      (0): Linear(in_features=2901, out_features=64, bias=True)
      (1): Tanh()
      (2): Linear(in_features=64, out_features=64, bias=True)
      (3): Tanh()
    )
  )
  (action_net): Linear(in_features=64, out_features=4, bias=True)
  (value_net): Linear(in_features=64, out_features=1, bias=True)

**Comment by araffin** [comment]
Hello,
I'm glad you found out your answer.
To add to it, you could also use a CNN policy for 2D matrices and pass `normalize_images=False` to the policy.
See https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/torch_layers.py#L77-L86

**Comment by beuricav** [comment]
Thank you very much for this hint I will give it a try.
